packages/react-mongo/react-mongo.ts

```diff
@@ -0,0 +1,63 @@
+import { Meteor } from 'meteor/meteor'
+import { Mongo } from 'meteor/mongo'
+import { Tracker } from 'meteor/tracker'
+import { useEffect, useMemo, useReducer, useRef, DependencyList } from 'react'
+
+const fur = (x: number): number => x + 1
+const useForceUpdate = () => useReducer(fur, 0)[1]
+
+const useSubscriptionClient = (factory: () => Meteor.SubscriptionHandle, deps: DependencyList = []) => {
+  const forceUpdate = useForceUpdate()
+  const subscription = useRef<Meteor.SubscriptionHandle>({
+    stop() {},
+    ready: () => false
+  })
+
+  useEffect(() => {
+    subscription.current = factory()
+    const computation = Tracker.autorun(() => {
+      if (subscription.current.ready()) forceUpdate()
+    })
```

Comment:
I figured this was desired behavior when using subscription, but are you saying there are cases where you  might not want to render on ready? I guess the cursor will actually take care of it - I never thought of that. I can definitely make that behavior optional.

Reply:
I think you'd only need that if you want to render a different loading state.

Reply:
What should the API for this look like? We already take deps as the second arg. If we consider this an advanced feature, we can take a config option instead of deps (with a deps prop). Would that be good enough? Or maybe we can pass an event like object to the factory, something with a `dontForceUpdate()` method or something.

Reply:
I came up with a neat way to do this. Basically, if you call the `handle.ready()` method, it'll note that, and then force update. If you don't, it won't. Easy peasy.

Reply:
Cool!

Reply:
Unfortunately, this triggers an error:

![image](https://user-images.githubusercontent.com/159500/97961864-5312ec80-1db4-11eb-904b-887845225a4d.png)

Reply:
While this idea is cool, it makes the code a lot harder to reason about IMO, with the fake handle that is swapped for a real handle, which I think also led to exactly this problem (attempting to call the `subscription.ready()` but calling the fake handle's `ready()` instead). Also, as the handle is stored in a ref, running the `useEffect` won't change the return value and you'll still get the fake handle even though the real one might already been stored in the `refs`. I think this needs some more work. Happy to do a PR to suggest some changes if you're ok with it.

Reply:
PR: #304

Reply:
The purpose of useSubscription is to set up the subscription, the auto unsubscribe/re-subscribe on deps change, the auto-unsubscribe on unmount (which the current code does right). That's enough value in itself, I'm not sure it should handle reactivity on ready stat top of that : we already have a tool to run reactive funcs, and that's useTracker :

const ready = useSubscription(...);
const isReady = useTracker(ready);

(I'm not convinced there's a real use case for useSubscription to return the full subsciption handle with stop() and all. Returning the ready() func ready for you to react to ready state change could be enough and simpler ?)

Reply:
Another issue I'm running into with this: when the `deps` change (and thus a new subscription is started), for the first render cycle the hook will still return the previous subscription's status (and so will return an invalid `ready() === true` for the first cycle). I will try to fix this in the PR as well.

Reply:
Also : is there a real value in passing a factory function doing the subscribe call (but maybe also doing / returning something bogus ) ?
Why not make it a an actual wrapper / replacement for Meteor.subscribe, just like useUser() wraps / hides Meteor.user() ?

useSubscription('mySubsname', ...args), and then no need for a deps arg ?
or useSubscription(someCondition && ['mySubsname', ...args]) for the conditionally subscribe form ?

Reply:
I have been thinking about this as well, and agree with @yched . It actually looks a lot better, and makes the hook way more simple to implement internally as well.

Reply:
I am writing a different version now that integrates with [`use-asset`](https://github.com/pmndrs/use-asset) to support `Suspense`. It's actually really easy and you can do things like this:

```tsx
function Board() {
  return (
    <div className="board">
      <Suspense fallback={<b>Loading…</b>}>
        <Projects boardId={x} />
      </Suspense>
    </div>
  )
}

function Projects({ boardId }) {
  useSubscription( !!boardId && 'board.projects', boardId )
  
  // No need for .ready() checking; when we get here we know it's ready
  
  const projects = useCursor( () => { /* ... */ } )
}
```

In my early testing it is working really well, and most importantly solves [the issue I am having with the other hook](https://github.com/meteor/react-packages/pull/298#discussion_r516527356). I also really enjoy this style of coding.

Let me know if you want me to make a PR for this package. If not, I'll maybe release it as a separate package.

Reply:
In addition, `Suspense` allows you do to [all kinds of crazy stuff](https://codesandbox.io/s/use-asset-dependency-70908) like waiting to load non-essential data until the core data has loaded:

```tsx
function Board() {
  return (
    <div className="board">
      <Suspense fallback={<b>Loading…</b>}>
        <Projects {...{ boardId, order, group }} />
        
        {/* This will only be rendered once <Projects> no longer suspends */}
        <LoadMenuData />
      </Suspense>
    </div>
  )
}

function LoadMenuData() {
  useSubscriptionAsset( 'boards' )
  return null
}
```

Reply:
@CaptainN @yched FYI, I put up a package with this concept, more info here: https://forums.meteor.com/t/usesubscription-with-react-suspense-support/54439

Reply:
That's awesome! Suspense for data is not officially supported by React yet, and SSR is still tricky (last time I checked anyway), but these are definitely things I'd love to see in an official Meteor hook.